### DISCUSSION TITLE
DONE: Option to lock some sheet icons

### DISCUSSION THREAD

**Discussion by tm-henningnt** [ptarmiganlabs/butler-sheet-icons#197] [Ideas]
Would be great if there was a way to lock icons for certain sheets. In many cases we want custom icons for a few sheets in an app, and then auto generated thumbnails for the others.

**Comment by mountaindude** [comment]
Great idea!

Somehow we must then identify which sheets should not be updated by BSI.
I can see several options for this:

1. Specify one or more sheet name(s) that BSI will ignore. Probably the easiest to use, but also a theoretical risk of false positives.
2. Specify exclude sheets using sheet IDs. No risk of false positives but sheet IDs would have to be copy-pasted from the app's URL. Or another command could be added to BSI that lists sheet names and IDs for some/all apps.
3. Add option that lets the user say "exclude sheet 'n' in all apps", with 'n' being the position of the sheet in the app.
4. ...something else?

In the spirit of keeping things simple to use I am leaning towards option 1 and possibly also 3. 

Thoughts?

**Reply by tm-henningnt** [reply]
How about tags or custom properties on the sheet? Or skip sheets which have sheet icons set from a content library?

**Comment by mountaindude** [comment]
Good thinking, but sheets can't have tags or custom properties.

And... After running BSI *all* sheets will have an icons from a content library. 
But sure, it would be good to have an option to only create sheet icons for sheets that do not already have a sheet icon.
Somewhat limited use case but could still be useful.

Another idea (already tracked [here](https://github.com/ptarmiganlabs/butler-sheet-icons/discussions/198)) is to add a command for removing sheet icons from app(s).

**Comment by tm-henningnt** [comment]
Looks like you can have tags? 
![image](https://user-images.githubusercontent.com/73784875/159935320-8d17b73f-00e2-4b75-a7da-860aa7a7c3de.png)

**Reply by mountaindude** [reply]
You are absolutely right.
Sheets are app objects. App objects can have tags but not custom properties (I think...).
I think we have a winner there then.


Tag the sheets that should NOT be updated.
or 
Specify which sheet number within each app should NOT be updated.
or
Specify which sheet names should NOT be updated

That should cover most needs I think.

**Comment by mountaindude** [comment]
# Design notes

Sheet excluding can be done in several ways. 
These are additive options, meaning that sheets matching _any_ of the options will be excluded.

## Client-managed Qlik Sense Enterprise on Windows

Three ways to control which sheets' icons should NOT be updated:

1. Tag the sheets that should NOT be updated. `--exclude-sheet-tag`. Exactly one tag name can be specified.
2. Sheet number within each app should NOT be updated. `--exclude-sheet-number`. One or more numbers representing sheet position within app. 1-based, i.e. 1 is first sheet. 
3. Which sheet names should NOT be updated. `--exclude-sheet-title`. One or more text strings can be specified. Exact match against the complete sheet title will be done.

## Qlik Sense Cloud

Same as for QSEoW except that tag-based sheet selection is not possible in QS Cloud. 
Option 2 and 3 remain.
